### PR TITLE
Remove usage of encoding.MustDecode* routines.

### LINKED ISSUE #2671: sql: improve error handling when decoding table keys
`sql.decodeTableKey` currently blindly assumes that the encoded data matches the expected column type. If the expectation doesn't match we can get panics in the decoding routines. We're not currently and haven't historically seen panics from this code, but it is probably worthwhile to be defensive. 

We could change the decoding routines to return an error instead of panicing, though that seems onerous to the callers. We could add some sort of verification routine that can be called first, though that also places a burden on the callers to remember to call that routine. On the other hand, the decoding routines are not called from that many places (at least in the `sql` package) so perhaps any worry about additional burden is misplaced.

### PR DESCRIPTION
**Pull request by petermattis**
Propagate decoding errors everywhere except from keys.KeyAddress().

Fixes #2671.

### CODE REVIEW

File: ts/keys.go [hunk 1]
```diff
@@ -78,31 +78,33 @@ func MakeDataKey(name string, source string, r Resolution, timestamp int64) prot
...
-func DecodeDataKey(key proto.Key) (string, string, Resolution, int64) {
-	var (
-		name          []byte
-		source        []byte
-		resolutionInt int64
-		timeslot      int64
-		remainder     = key
-	)
-
+func DecodeDataKey(key proto.Key) (string, string, Resolution, int64, error) {
```
Comment (vivekmenezes) [inline ts/keys.go]: This might be a good case for named return parameters.

File: storage/engine/mvcc.go [hunk 2]
```diff
@@ -1662,18 +1695,28 @@ func mvccEncodeTimestamp(key proto.EncodedKey, timestamp proto.Timestamp) proto.
 // exactly 12 trailing bytes and they're decoded into a timestamp.
 // The decoded key, timestamp and true are returned to indicate the
 // key is for an MVCC versioned value.
-func MVCCDecodeKey(encodedKey proto.EncodedKey) (proto.Key, proto.Timestamp, bool) {
-	tsBytes, key := encoding.MustDecodeBytes(encodedKey, nil)
+func MVCCDecodeKey(encodedKey proto.EncodedKey) (proto.Key, proto.Timestamp, bool, error) {
```
Comment (vivekmenezes) [inline storage/engine/mvcc.go]: same here. Perhaps using named return parameters is the way to go.

File: storage/gc_queue_test.go [hunk 3]
```diff
@@ -349,7 +355,9 @@ func TestGCQueueIntentResolution(t *testing.T) {
 	// Iterate through all values to ensure intents have been fully resolved.
 	meta := &engine.MVCCMetadata{}
 	err := tc.store.Engine().Iterate(engine.MVCCEncodeKey(proto.KeyMin), engine.MVCCEncodeKey(proto.KeyMax), func(kv proto.RawKeyValue) (bool, error) {
-		if key, _, isValue := engine.MVCCDecodeKey(kv.Key); !isValue {
+		if key, _, isValue, err := engine.MVCCDecodeKey(kv.Key); err != nil {
+			t.Fatal(err)
```
Comment (tschottdorf) [inline storage/gc_queue_test.go]: in this function, looks like you could just return an error.

**Comment by vivekmenezes** [comment]
LGTM

File: storage/gc_queue_test.go (hunk 3 above)
Comment (tschottdorf) [inline storage/gc_queue_test.go]: also below in previous code.

**Comment by tschottdorf** [comment]
LGTM. I can kind of see the point in named returns where @vivekmenezes suggests it, but if the comment explains it well, maybe not.

File: ts/keys.go (hunk 1 above)
Comment (petermattis) [inline ts/keys.go]: Maybe, though named return parameters have their own downsides. I'm just passing by this code and will let @mrtracy fix this up in a subsequent PR if he wishes.

File: storage/engine/mvcc.go (hunk 2 above)
Comment (petermattis) [inline storage/engine/mvcc.go]: I'm not sure that would clarify anything. Yes, there are 4 return values, but their types are all distinct. If you get the order confused the compiler will warn you. If you leave a return value out the compiler will warn you.

File: storage/gc_queue_test.go (hunk 3 above)
Comment (petermattis) [inline storage/gc_queue_test.go]: Yeah. Not sure if that buys us much (same number of lines of code), but it is done.

File: ts/keys.go (hunk 1 above)
Comment (petermattis) [inline ts/keys.go]: FWIW, I don't think the returned `error` is a good reason to use named returns. But the previous returned values are (i.e. what is the difference between the 2 strings).

File: keys/keys.go [hunk 4]
```diff
@@ -151,17 +140,20 @@ func MakeRangeKey(key, suffix, detail proto.Key) proto.Key {
 
 // DecodeRangeKey decodes the range key into range start key,
 // suffix and optional detail (may be nil).
-func DecodeRangeKey(key proto.Key) (startKey, suffix, detail proto.Key) {
+func DecodeRangeKey(key proto.Key) (startKey, suffix, detail proto.Key, err error) {
 	if !bytes.HasPrefix(key, LocalRangePrefix) {
-		panic(fmt.Sprintf("key %q does not have %q prefix",
-			key, LocalRangePrefix))
+		return nil, nil, nil, fmt.Errorf("key %q does not have %q prefix",
```
Comment (tamird) [inline keys/keys.go]: util?

File: storage/engine/gc.go [hunk 5]
```diff
@@ -88,7 +92,11 @@ func (gc *GarbageCollector) Filter(keys []proto.EncodedKey, values [][]byte) pro
 	// If there are no non-deleted survivors, return timestamp of first key
 	// to delete all entries.
 	if !survivors {
-		_, ts, _ := MVCCDecodeKey(keys[0])
+		_, ts, _, err := MVCCDecodeKey(keys[0])
+		if err != nil {
+			log.Errorf("unable to decode MVCC key: %q: %v", keys[0], err)
```
Comment (tamird) [inline storage/engine/gc.go]: can you add something to this log line that indicates that all entries will be deleted?

File: keys/keys.go (hunk 4 above)
Comment (petermattis) [inline keys/keys.go]: Done. My fingers do not want to type `util.Errorf`.

File: storage/engine/gc.go (hunk 5 above)
Comment (tschottdorf) [inline storage/engine/gc.go]: `proto.ZeroTimestamp` indicates that nothing will be deleted, unless there's a caveat I'm missing.

Comment (petermattis) [inline storage/engine/gc.go]: Sure.

Comment (tamird) [inline storage/engine/gc.go]: @tschottdorf that seems to be in conflict with the comment above this block?

Comment (petermattis) [inline storage/engine/gc.go]: Oh, @tschottdorf is right. I misread that just now, but the comment and usage are clear.

Comment (tschottdorf) [inline storage/engine/gc.go]: the comment is accurate, but only if the returned timestamp is not `proto.ZeroTimestamp` (at least if the callers heed the documentation of this function at the top).

Comment (tamird) [inline storage/engine/gc.go]: Oh, then, if we can't decode the key we should return the smallest non-zero timestamp, yeah?

Comment (tschottdorf) [inline storage/engine/gc.go]: if you can't decode it, I'm not sure you want to gc everything. Better to leave it in place for eventual inspection - but of course there's some value in reporting this up (potential ReplicaCorruptionError). Didn't we have a crawler queue whose sole purpose it is to touch everything?

Comment (tamird) [inline storage/engine/gc.go]: +1 ReplicaCorruptionError

Comment (tamird) [inline storage/engine/gc.go]: @petermattis can you add a TODO here to propagate this error up?

Comment (petermattis) [inline storage/engine/gc.go]: Added a `TODO(tschottdorf)`. :)

**Pull request merged**